packages/graph/hash_graph/lib/graph/src/store/postgres/migration.rs

```diff
@@ -0,0 +1,81 @@
+use async_trait::async_trait;
+use error_stack::{IntoReport, Result, ResultExt};
+use tokio_postgres::Client;
+
+use super::{AsClient, PostgresStore};
+use crate::store::{
+    error::MigrationError,
+    migration::{Migration, MigrationState, StoreMigration},
+};
+
+mod embedded {
+    use refinery::embed_migrations;
+    embed_migrations!("../../postgres_migrations");
+}
+
+#[async_trait]
+impl<C: AsClient<Client = Client>> StoreMigration for PostgresStore<C> {
+    async fn run_migrations(&mut self) -> Result<(), MigrationError> {
+        embedded::migrations::runner()
+            .run_async(self.as_mut_client())
+            .await
+            .into_report()
+            .change_context(MigrationError)?;
+
+        Ok(())
+    }
+
+    async fn all_migrations(&mut self) -> Result<Vec<Migration>, MigrationError> {
+        let all_migrations = embedded::migrations::runner()
+            .get_migrations()
+            .iter()
+            .map(|migration| migration.into())
+            .collect();
+
+        Ok(all_migrations)
+    }
+
+    async fn applied_migrations(&mut self) -> Result<Vec<Migration>, MigrationError> {
+        let applied_migrations = embedded::migrations::runner()
+            .get_applied_migrations_async(self.as_mut_client())
+            .await
+            .into_report()
+            .change_context(MigrationError)?
+            .iter()
+            .map(|migration| migration.into())
+            .collect();
+
+        Ok(applied_migrations)
+    }
+
+    async fn missing_migrations(&mut self) -> Result<Vec<Migration>, MigrationError> {
+        let all_migrations = self.all_migrations().await?;
+        let applied_migrations = self.all_migrations().await?;
+
+        // Migrations are expected to be a very small list, even with thousands of migrations, the
+        // performance implications of this
+        let difference: Vec<_> = all_migrations
+            .into_iter()
+            .filter(|item| !applied_migrations.contains(item))
+            .collect();
```

Comment:
Too bad that `drain_filter` is unstable, but it's not worth adding a feature for that:
```rust
let difference: Vec<_> = all_migrations
    .drain_filter(|m| !applied_migrations.contains(m))
    .collect();
```

Reply:
Yeah, would be cleaner, but not worth it if we don't require that feature elsewhere